### PR TITLE
Linalg changes from oed

### PR DESCRIPTION
**Pull request by ae-foster**
Changes:
 - new `rinverse` method for batch inversion of matrices (on rightmost 2 dims)
 - pytorch helper functions for OED work
 - some tests

### CODE REVIEW

File: pyro/ops/newton.py [hunk 1]
```diff
@@ -1,66 +1,10 @@
...
-
-    Hinv[..., 0, 1] = Hinv01
-    Hinv[..., 1, 0] = Hinv01
-    Hinv[..., 0, 2] = Hinv02
-    Hinv[..., 2, 0] = Hinv02
-    Hinv[..., 1, 2] = Hinv12
-    Hinv[..., 2, 1] = Hinv12
-    Hinv = Hinv / detH.unsqueeze(-1).unsqueeze(-1)
-    return Hinv
+from pyro.ops.linalg import rinverse, _eig_3d
```
Comment (fritzo) [inline pyro/ops/newton.py]: Functions that are imported across files should not start with underscore; could you rename to `eig_3d`?

File: pyro/ops/linalg.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        inv[..., :d, d:] = -Ainv.matmul(B).matmul(schur)
+        if sym:
+            inv[..., d:, :d] = inv[..., :d, d:].transpose(-1, -2)
+        else:
+            inv[..., d:, :d] = -schur.matmul(C).matmul(Ainv)
+        inv[..., d:, d:] = schur
+        return inv
+
+
+def _determinant_3d(H):
```
Comment (fritzo) [inline pyro/ops/linalg.py]: rename to `determinant_3d`

File: pyro/ops/linalg.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    """
+    Returns the determinants of a batched 3-D matrix
+    """
+    detH = (H[..., 0, 0] * (H[..., 1, 1] * H[..., 2, 2] - H[..., 2, 1] * H[..., 1, 2]) +
+            H[..., 0, 1] * (H[..., 1, 2] * H[..., 2, 0] - H[..., 1, 0] * H[..., 2, 2]) +
+            H[..., 0, 2] * (H[..., 1, 0] * H[..., 2, 1] - H[..., 2, 0] * H[..., 1, 1]))
+    return detH
+
+
+def _eig_3d(H):
```
Comment (fritzo) [inline pyro/ops/linalg.py]: rename to `eig_3d`

File: pyro/ops/linalg.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    phi[r > 1 - 1e-6] = 0.
+
+    eig1 = q + 2 * p * torch.cos(phi[..., 0, 0])
+    eig2 = q + 2 * p * torch.cos(phi[..., 0, 0] + (2 * math.pi/3))
+    eig3 = 3 * q - eig1 - eig2
+    # eig2 <= eig3 <= eig1
+    return eig2, eig3, eig1
+
+
+def _inv_3d(H, sym=False):
```
Comment (fritzo) [inline pyro/ops/linalg.py]: rename to `inv_3d`

**Comment by ae-foster** [comment]
Should be all sorted now @fritzo

**Pull request merged**